### PR TITLE
Improve http retry

### PR DESCRIPTION
**Pull request by samansmink**
this PR aims makes the https retry mechanism consistent between request types and configurable:

Each HEAD, GET, PUT, and POST request will now be reattempted on:
- socket read/write error
- socket connection error
- various HTTP errors such as 429 (rate limit hit) and 503/504 (server error/timeout)

Retry mechanism uses retry wait times with configurable exponential backoff from the second retry, configurable through:
- http_retries: total number of retries
- http_retry_wait_ms: base wait time before retrying
- http_retry_backoff: backoff factor for exponential backoff

defaults are (http_retries: 3, http_retry_wait_ms: 100, http_retry_backoff:4) which results in this behaviour:
initial try -> wait 0ms -> retry1 -> wait 100ms -> retry2 -> wait 400ms -> (retry 3) -> report IOException

note that the first retry is always done without waiting.

This means that a request to a non-existing file will wait a total amount of 500ms which seems like a reasonable balance between responsiveness and robustness. 

For large uploads/downloads it may be useful to use more requests and/or longer waits.

### CODE REVIEW

File: extension/httpfs/httpfs.cpp [hunk 1]
```diff
@@ -30,23 +36,86 @@ HTTPParams HTTPParams::ReadFrom(FileOpener *opener) {
...
+		if (should_retry && tries <= params.retries) {
+			if (tries > 1) {
+				uint64_t sleep_amount = (uint64_t)((float)params.retry_wait_ms * pow(params.retry_backoff, tries - 2));
+				std::this_thread::sleep_for(std::chrono::milliseconds(sleep_amount));
+			}
+			if (retry_cb) {
+				retry_cb();
+			}
+		} else {
+			throw IOException("HTTP " + method + " error on '" + url + "' (Error code " + to_string((int)res.error()) +
```
Comment (Mause) [inline extension/httpfs/httpfs.cpp:116]: If you remove the int cast here, we'll get the actual error code name instead (if the `to_string` overload is in scope)

File: extension/httpfs/httpfs.cpp [hunk 2]
```diff
@@ -349,14 +406,14 @@ unique_ptr<ResponseWrapper> HTTPFileHandle::Initialize() {
 	if (res->code != 200) {
 		if ((flags & FileFlags::FILE_FLAGS_WRITE) && res->code == 404) {
 			if (!(flags & FileFlags::FILE_FLAGS_FILE_CREATE) && !(flags & FileFlags::FILE_FLAGS_FILE_CREATE_NEW)) {
-				throw std::runtime_error("Unable to open URL \"" + path +
-				                         "\" for writing: file does not exists and CREATE flag is not set");
+				throw IOException("Unable to open URL \"" + path +
+				                  "\" for writing: file does not exists and CREATE flag is not set");
```
Comment (Mause) [inline extension/httpfs/httpfs.cpp:410]:
```suggestion
				                  "\" for writing: file does not exist and CREATE flag is not set");
```

File: extension/httpfs/s3fs.cpp [hunk 3]
```diff
@@ -265,49 +265,17 @@ void S3FileSystem::UploadBuffer(S3FileHandle &file_handle, shared_ptr<S3WriteBuf
...
-		                         last_error + "\")");
+	if (res->code != 200) {
+		throw IOException("Unable to connect to URL " + file_handle.path + " " + res->error + " (HTTP code " +
+		                  to_string(res->code) + ")");
 	}
 
 	auto etag_lookup = res->headers.find("ETag");
 	if (etag_lookup == res->headers.end()) {
-		throw std::runtime_error("Unexpected reponse when uploading part to S3");
+		throw IOException("Unexpected reponse when uploading part to S3");
```
Comment (Mause) [inline extension/httpfs/s3fs.cpp:278]:
```suggestion
		throw IOException("Unexpected response when uploading part to S3");
```

File: extension/httpfs/s3fs.cpp [hunk 4]
```diff
@@ -946,8 +913,8 @@ string AWSListObjectV2::Request(string &path, HTTPParams &http_params, S3AuthPar
 		    return true;
 	    });
 	if (res.error() != duckdb_httplib_openssl::Error::Success) {
-		throw std::runtime_error("HTTP GET error on '" + listobjectv2_url + "' (Error code " +
-		                         to_string((int)res.error()) + ")");
+		throw IOException("HTTP GET error on '" + listobjectv2_url + "' (Error code " + to_string((int)res.error()) +
```
Comment (Mause) [inline extension/httpfs/s3fs.cpp:916]: Ditto for removing int cast

**Review by Mause — CHANGES_REQUESTED** [review]

File: extension/httpfs/httpfs.cpp (hunk 1 above)
Comment (samansmink) [reply extension/httpfs/httpfs.cpp]: good point, thats much nicer!

**Comment by Mytherin** [comment]
Thanks! LGTM

**Pull request merged**